<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Launch Sites Locations Analysis with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:
- **TASK 1:** Mark all launch sites on a map
- **TASK 2:** Mark the success/failed launches for each site on the map
- **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [2]:
! pip3 install folium
! pip3 install wget
! pip3 install pandas

  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9711 sha256=b9b1786e7fe1140dcafcce5f26acade2cb4d9c1ed96be3be7b7a0ce0bb3eaec0
  Stored in directory: c:\users\infer\appdata\local\pip\cache\wheels\40\b3\0f\a40dbd1c6861731779f62cc4babcb234387e11d697df70ee97
Successfully built wget


In [1]:
import folium
import wget
import pandas as pd

In [2]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/DV0101EN-3-5-1-Generating-Maps-in-Python-py-v2.0.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site. 


In [3]:
# Download and read the `spacex_launch_geo.csv`
spacex_csv_file = wget.download('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv')
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [4]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [5]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example, 


In [7]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle. 


Now, let's add a circle for each launch site in data frame `launch_sites`


_TODO:_  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [8]:
# Initial the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=6)
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label
ccafs_lc_coord = [launch_sites_df.loc[launch_sites_df['Launch Site'] == 'CCAFS LC-40', 'Lat'].values[0],
               launch_sites_df.loc[launch_sites_df['Launch Site'] == 'CCAFS LC-40', 'Long'].values[0]]

ccafs_slc_coord = [launch_sites_df.loc[launch_sites_df["Launch Site"] == "CCAFS SLC-40", "Lat"].values[0],
                   launch_sites_df.loc[launch_sites_df["Launch Site"] == "CCAFS SLC-40", "Long"].values[0]]

ksc_coord = [launch_sites_df.loc[launch_sites_df["Launch Site"] == "KSC LC-39A", "Lat"].values[0],
            launch_sites_df.loc[launch_sites_df["Launch Site"] == "KSC LC-39A", "Long"].values[0]]

vafb_coord = [launch_sites_df.loc[launch_sites_df["Launch Site"] == "VAFB SLC-4E", "Lat"].values[0],
              launch_sites_df.loc[launch_sites_df["Launch Site"] == "VAFB SLC-4E", "Long"].values[0]]

#Agregar circulos
# Círculo
folium.Circle(
    ccafs_lc_coord,
    radius=1000,
    color='#d35400',
    fill=True
).add_child(
    folium.Popup('CCAFS LC-40')
).add_to(site_map)

# Etiqueta
folium.map.Marker(
    ccafs_lc_coord,
    icon=DivIcon(
        icon_size=(20, 20),
        icon_anchor=(0, 0),
        html='<div style="font-size: 12px; color:#d35400;"><b>CCAFS LC-40</b></div>'
    )
).add_to(site_map)

#ccafs_slc
folium.Circle(ccafs_slc_coord, radius=1000, color='blue', fill=True).add_child(
    folium.Popup("CCAFS SLC-40")
).add_to(site_map)

folium.map.Marker(ccafs_slc_coord, icon=DivIcon(
    icon_size=(20, 20), icon_anchor=(0, 0),
    html='<div style="font-size: 12px; color: blue;"><b>CCAFS SLC-40</b></div>')
).add_to(site_map)


#ksc
folium.Circle(ksc_coord, radius=1000, color='green', fill=True).add_child(
    folium.Popup("KSC LC-39A")
).add_to(site_map)
folium.map.Marker(ksc_coord, icon=DivIcon(
    icon_size=(20,20), icon_anchor=(0,0),
    html='<div style="font-size: 12px; color: green;"><b>KSC LC-39A</b></div>')
).add_to(site_map)

#VAFB
folium.Circle(vafb_coord, radius=1000, color="purple", fill=True).add_child(
    folium.Popup("VAFB SLC-4E")
).add_to(site_map)
folium.map.Marker(vafb_coord, icon=DivIcon(
    icon_size=(20,20), icon_anchor=(0,0),
    html='<div style="font-size: 12px; color: purple;"><b>VAFB SLC-4E</b></div>')
).add_to(site_map)
site_map

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:
- Are all launch sites in proximity to the Equator line?
- Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


### 1.- Launch sites are relatively far from the Ecuator line.
### 2.- Yes, launch sites are close to the coast

# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [9]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records. 
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [10]:
marker_cluster = MarkerCluster().add_to(site_map)

for lat, lon, outcome in zip(spacex_df["Lat"], spacex_df["Long"], spacex_df["class"]):
    color = "green" if outcome == 1 else "red"
    folium.Marker(
        location=[lat,lon],
        icon=folium.Icon(color=color)
    ).add_to(marker_cluster)

site_map

_TODO:_ Create a new column in `launch_sites` dataframe called `marker_color` to store the marker colors based on the `class` value


In [11]:

# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red
def assign_marker_color(launch_outcome):
    if launch_outcome == 1:
        return "green"
    else:
        return "red"

spacex_df["marker_color"] = spacex_df["class"].apply(assign_marker_color)
spacex_df.sample(5)

Launch Site        Lat       Long  class marker_color
17   CCAFS LC-40  28.562302 -80.577356      1        green
48    KSC LC-39A  28.573255 -80.646895      1        green
38    KSC LC-39A  28.573255 -80.646895      1        green
55  CCAFS SLC-40  28.563197 -80.576820      0          red
8    CCAFS LC-40  28.562302 -80.577356      0          red

_TODO:_ For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [12]:
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, 
# e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
for index, record in spacex_df.iterrows():
   marker = folium.Marker(
     location=[record["Lat"], record["Long"]],
     icon=folium.Icon(color=record["marker_color"])
   )
   marker_cluster.add_child(marker)

site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [13]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


You can calculate the distance between two points on the map based on their `Lat` and `Long` values using the following method:


In [14]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

_TODO:_ Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [24]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)
#desempaquetar coordenadas VAFB
vafb_lat, vafb_lon = vafb_coord

west_coast_lat = 34.63702 
west_coast_lon = -120.62422

distance_coastline = calculate_distance(vafb_lat, vafb_lon, west_coast_lat, west_coast_lon)
print(f"Distancia a la costa: {distance_coastline:.2f} km")

#marcar en el mapa
folium.Marker(
    location=[west_coast_lat, west_coast_lon],
    icon=folium.Icon(color="blue"),
    popup=f"Distance to Launch Site: {distance_coastline:.2f} km"
).add_to(site_map)

site_map

Distancia a la costa: 1.32 km


_TODO:_ After obtained its coordinate, create a `folium.Marker` to show the distance


In [29]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
# distance_marker = folium.Marker(
#    coordinate,
#    icon=DivIcon(
#        icon_size=(20,20),
#        icon_anchor=(0,0),
#        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance),
#        )
#    )

folium.Marker(
    location=[west_coast_lat, west_coast_lon],
    icon=DivIcon(
        icon_size=(100, 20),
        icon_anchor=(0, 0),
        html=f'<div style="font-size: 12px; color: #d35400;"><b>{distance_coastline:.2f} KM</b></div>',
    )
).add_to(site_map)



_TODO:_ Draw a `PolyLine` between a launch site to the selected coastline point


In [30]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
# lines=folium.PolyLine(locations=coordinates, weight=1)
folium.PolyLine(
    locations=[[vafb_lat, vafb_lon], [west_coast_lat, west_coast_lon]],
    color="blue",
    weight=2,
    dash_array="5,5"
).add_to(site_map)
site_map

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


_TODO:_ Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [39]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site
ksc_lat, ksc_lon = ksc_coord

titus_lat = 28.61234
titus_lon = -80.80918

#calcular distancia 
titus_distance = calculate_distance(ksc_lat,ksc_lon, titus_lat, titus_lon)
print(f"Distancia de base KSC-LC 39A a Titusville: {titus_distance:.2f} km")

#Etiquetar 
folium.Marker(
    location=[titus_lat,titus_lon],
    icon=DivIcon(
        icon_size=(100,20),
        icon_anchor=(0,0),
        html=f'<div style="font-size: 12px; color="purple";><b>{titus_distance:.2f} KM</b></div>', 
    )
).add_to(site_map)
#dibujar linea desde Titusville a base KSC
folium.PolyLine(
    locations=[[ksc_lat, ksc_lon], [titus_lat, titus_lon]],
    color="purple",
    weight=2,
    dash_array="5,5"
).add_to(site_map)


Distancia de base KSC-LC 39A a Titusville: 16.43 km


In [48]:
ccsafs_lat, ccafs_lon = ccafs_lc_coord

railway_lat = 28.57967
railway_lon = -80.80516

#calcular distancia 
railway_distance = calculate_distance(ccsafs_lat, ccafs_lon, railway_lat, railway_lon)
print(f"Distancia de base CCSAFS-40 a via de tren: {titus_distance:.2f} km")

#etiquetar en mapa
folium.Marker(
    location=[railway_lat, railway_lon],
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html=f'<div style="font-size: 12px; color=red";><b>{railway_distance:.2f} KM</b></div>',
    )
).add_to(site_map)

#Linea 
folium.PolyLine(
    locations=[[ccsafs_lat,ccafs_lon], [railway_lat,railway_lon]],
    color="red",
    weight=2,
    dash_array="5,5"
).add_to(site_map)


Distancia de base CCSAFS-40 a via de tren: 16.43 km


In [53]:
#desempaquetando ubicacion
ccafs_slc_lat, ccafs_slc_lon =ccafs_slc_coord

#lat,lon autopista Nasa
nasa_parkway_lat = 28.49559
nasa_parkway_lon = -80.58431

#calcular distancia entre CCAFS_SLC a Nasa parkway

nasa_highway_distance = calculate_distance(ccafs_slc_lat, ccafs_slc_lon, nasa_parkway_lat,nasa_parkway_lon)
print(f"Distancia de base CCAFS SLC 40 a autopista Nasa Parkway: {nasa_highway_distance:.2f} km")

#etiquetar lugar
folium.Marker(
    location=[nasa_parkway_lat, nasa_parkway_lon],
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html=f'<div style="font-size: 12px; color:gray;"><b>{nasa_highway_distance:.2f} KM</b></div>',
    )
).add_to(site_map)
#linea
folium.PolyLine(
    locations=[[ccafs_slc_lat,ccafs_slc_lon], [nasa_parkway_lat, nasa_parkway_lon]],
    color="gray",
    weight=2,
    dash_array="5,5"
).add_to(site_map)
site_map

Distancia de base CCAFS SLC 40 a autopista Nasa Parkway: 7.56 km


After you plot distance lines to the proximities, you can answer the following questions easily:
- Are launch sites in close proximity to railways?
- Are launch sites in close proximity to highways?
- Are launch sites in close proximity to coastline?
- Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


1. Launch sites are not very close to railways.  
2. Launch sites are closer to some highways, probably due to logistical needs.  
3. Yes, they are in close proximity to the coastline, likely to minimize damage risks in case of a launch failure.  
4. Yes, in the case of the East Coast, the nearest city (Titusville) is 16.43 km away, indicating a safe distance from urban areas.


# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Yan Luo](https://www.linkedin.com/in/yan-luo-96288783/)


### Other Contributors


Joseph Santarcangelo


## Change Log


|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2021-05-26|1.0|Yan|Created the initial version|


Copyright © 2021 IBM Corporation. All rights reserved.
